File: python-backend/app/models/conversation.py
"""Conversation data models."""

from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import uuid4

from pydantic import BaseModel, Field

from app.models.message import Message


class ConversationType(str, Enum):
    """Type of conversation."""
    SINGLE = "single"      # Single agent conversation
    DELEGATOR = "delegator"  # Main agent with sub-agents
    GROUP = "group"        # Multiple independent agents


class Conversation(BaseModel):
    """A conversation with one or more agents."""
    id: str = Field(default_factory=lambda: str(uuid4()))
    title: Optional[str] = None
    type: ConversationType
    agent_ids: list[str] = Field(default_factory=list)
    skill_ids: list[str] = Field(default_factory=list)
    messages: list[Message] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    # Typing state: agent_id -> is_typing
    typing_agents: dict[str, bool] = Field(default_factory=dict)
    
    # Metadata
    is_archived: bool = False
    
    class Config:
        use_enum_values = True


class ConversationCreate(BaseModel):
    """Request to create a new conversation."""
    type: ConversationType
    title: Optional[str] = None
    agent_ids: list[str]
    skill_ids: list[str] = Field(default_factory=list)
    
    class Config:
        use_enum_values = True


class ConversationResponse(BaseModel):
    """Response containing conversation data."""
    id: str
    title: Optional[str]
    type: str
    agent_ids: list[str]
    skill_ids: list[str]
    messages: list[Message]
    created_at: datetime
    updated_at: datetime
    typing_agents: dict[str, bool]
    is_archived: bool
    
    class Config:
        from_attributes = True


class ConversationSummary(BaseModel):
    """Summary of a conversation for list views."""
    id: str
    title: Optional[str]
    type: str
    agent_ids: list[str]
    skill_ids: list[str]
    last_message: Optional[str] = None
    last_message_time: Optional[datetime] = None
    unread_count: int = 0
    created_at: datetime
    updated_at: datetime

File: src/types/conversation.ts
/**
 * Conversation data types.
 */

import type { Message, MessageFromAPI } from './message';

/** Type of conversation */
export type ConversationType = 'single' | 'delegator' | 'group';

/** A conversation with one or more agents */
export interface Conversation {
  id: string;
  title: string | null;
  type: ConversationType;
  agentIds: string[];
  skillIds: string[];
  messages: Message[];
  createdAt: Date;
  updatedAt: Date;
  typingAgents: Record<string, boolean>;
  isArchived: boolean;
}

/** Request to create a new conversation */
export interface ConversationCreate {
  type: ConversationType;
  title?: string;
  agentIds: string[];
  skillIds?: string[];
}

/** Summary of a conversation for list views */
export interface ConversationSummary {
  id: string;
  title: string | null;
  type: ConversationType;
  agentIds: string[];
  skillIds: string[];
  lastMessage: string | null;
  lastMessageTime: Date | null;
  unreadCount: number;
  createdAt: Date;
  updatedAt: Date;
}

/** Conversation from API (snake_case) */
export interface ConversationFromAPI {
  id: string;
  title: string | null;
  type: ConversationType;
  agent_ids: string[];
  skill_ids: string[];
  messages: MessageFromAPI[];
  created_at: string;
  updated_at: string;
  typing_agents: Record<string, boolean>;
  is_archived: boolean;
}

/** Conversation summary from API */
export interface ConversationSummaryFromAPI {
  id: string;
  title: string | null;
  type: ConversationType;
  agent_ids: string[];
  skill_ids: string[];
  last_message: string | null;
  last_message_time: string | null;
  unread_count: number;
  created_at: string;
  updated_at: string;
}

/** Convert API conversation to frontend conversation */
export function parseConversation(api: ConversationFromAPI): Conversation {
  const { parseMessage } = require('./message');
  return {
    id: api.id,
    title: api.title,
    type: api.type,
    agentIds: api.agent_ids,
    skillIds: api.skill_ids,
    messages: api.messages.map(parseMessage),
    createdAt: new Date(api.created_at),
    updatedAt: new Date(api.updated_at),
    typingAgents: api.typing_agents,
    isArchived: api.is_archived,
  };
}

/** Convert API conversation summary to frontend */
export function parseConversationSummary(api: ConversationSummaryFromAPI): ConversationSummary {
  return {
    id: api.id,
    title: api.title,
    type: api.type,
    agentIds: api.agent_ids,
    skillIds: api.skill_ids,
    lastMessage: api.last_message,
    lastMessageTime: api.last_message_time ? new Date(api.last_message_time) : null,
    unreadCount: api.unread_count,
    createdAt: new Date(api.created_at),
    updatedAt: new Date(api.updated_at),
  };
}
